include/mxnet/buddy.h

```diff
@@ -0,0 +1,98 @@
+#ifndef MXNET_BUDDY_H_
+#define MXNET_BUDDY_H_
+
+#include <cuda_runtime_api.h>
+#include <iostream>
+#include <map>
+#include <math.h>
+#include <set>
+#include <stdio.h>
+#include <string>
+
+namespace mxnet {
+
+typedef enum {
+  blockStatus_Uninitialized,
+  blockStatus_Free,
+  blockStatus_Allocated
+} blockStatus_t;
+  
+const int MIN_ALLOC_SIZE = 128;
+const size_t CLEAN_UP_BOUNDRY = 500000000;
+
+inline int GetListIdx(size_t size) {
+  if (size <= 128) return 0;
+  return ceil(log2(static_cast<double>(size)) - log2(static_cast<double>(MIN_ALLOC_SIZE)));
+}
+
+inline unsigned long GetListSize(size_t size) {
+  return GetListIdx(size) + 1;
+}
+
+inline size_t GetListBlockSize(int idx) {
+  return pow(2, idx + log2(static_cast<double>(MIN_ALLOC_SIZE)));
```

Comment:
pow is very slow. I suggest using left shift instead

Reply:
I have not yet dived into buddy system. I'll do that in another branch. But I'll add a FIXME.